File: src/main/java/td/enastic/emploienastic/controler/StudentSocket.java
package td.enastic.emploienastic.controler;

import android.util.Log;
import td.enastic.emploienastic.model.*;

import java.io.*;
import java.net.*;
import java.util.ArrayList;
import java.util.List;

public class StudentSocket extends Thread implements SocketListener{

    private boolean connected;
    private LocalDatabase database;
    private LocalDatabase.Parametre setting;
    private EmploiDuTemp m_emploi;
    Socket socket = null;
    private  String objectName;
    BufferedReader inputStream = null;
    DataOutputStream outputStream = null;
    List<SocketListener> lsListeners;

    public StudentSocket(){
        lsListeners = new ArrayList<>();
        objectName = "";
        connected = false;
        socket = new Socket();
    }

    public void addListener(SocketListener l){
        lsListeners.add(l);
    }
    public void setDatabase(LocalDatabase database) {
        this.database = database;
        setting = database.getPrametre();
    }

    public void setSetting(LocalDatabase.Parametre setting) {
        this.setting = setting;
    }

    public LocalDatabase getDatabase() {
        return database;
    }

    public String getObjectName() {
        return objectName;
    }

    public void setObjectName(String objectName) {
        this.objectName = objectName;
    }

    public void tryToConnect() {
        if (socket!=null && socket.isConnected())
            return;
        while (true) {
            try {
                Thread.sleep(5000);
            } catch (InterruptedException e) {
                System.out.println(e.getMessage() + Thread.currentThread() + " tryToConnect");
                //e.printStackTrace();
            }
            try {
                socket = new Socket(setting.hostName(), (int) setting.portName());
                Log.i("MYDEBUG","Connexion reussu");
                inputStream = new BufferedReader(new InputStreamReader(socket.getInputStream()));
                outputStream = new DataOutputStream(socket.getOutputStream());
                status(Constant.ONLINE_ICON);
                onConnected();
                connected = true;
                break;
            } catch (UnknownHostException | SocketException | SocketTimeoutException e) {
                System.out.println(e.getMessage() + Thread.currentThread() + " tryToConnect");
                status(Constant.OFFLINE_ICON);
                connected = false;
                Log.i("MYDEBUG", "Error at StudentSocket.tryToconnect " + e.getMessage());
            } catch (IOException e) {
                connected = false;
                System.out.println(e.getMessage() + Thread.currentThread() + " tryToConnect");
                Log.i("MYDEBUG", "Error at StudentSocket.tryToconnect " + e.getMessage());
            }
        }
    }

    void onConnected(){
        if(!objectName.equals(Constant.MAIN_SOCKET_NAME))
            return;
        if(objectName.equals(Constant.MAIN_SOCKET_NAME)) {
            String data = Constant.FLAG + "\n" +
                    Constant.GET_LAST_EMPLOI_FLAG + "\n";
            writeUTF(data);
        }if(objectName.equals(Constant.ASK_TIMETABLE_SOCKET_NAME)){
            String data = Constant.FLAG+"\n"+
                    Constant.GET_SIGNATURE_EMPLOI_FLAG+"\n"+
                    m_emploi.toString();
            writeUTF(data);
        }
    }

    public void writeUTF(String data){
        try {
            outputStream.writeUTF(data);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
    public void onReadyRead() {

        int nbreEmbloi =0;
        while(true) {
            String flag = null;
            flag = readLine();
            flag = readLine();
            if (flag.equals(Constant.EMPLOI_FLAG)) {
                EmploiDuTemp em = readEmploi();
                //Log.i("MYDEBUG", "Emploi du temps recu: " + em.toString());
                if (1<0 && database.addEmploi(em)) {
                    //Demande de signature d'emploi
                    EmploiDuTemp tmpEmploi = new EmploiDuTemp(em.niveau(), "");
                    tmpEmploi.setNomSalle(em.nomSalle());
                    tmpEmploi.setDateDebut(em.dateDebut());
                    tmpEmploi.setLsHoraire(new ArrayList<>());
                    tmpEmploi.setLsMatieres(new ArrayList<>());
                    //askSignatureEmploi(em);
                    //sleep(1000);
                    nbreEmbloi++;
                }
            } else if (flag.equals(Constant.LAST_EMPLOI_FLAG)) {
                nbreEmbloi++;
                EmploiDuTemp em = readEmploi();
                //Log.i("MYDEBUG", "Emploi du temps recu: " + em.description());
                if (em.lsMatieres().size() > 0) {
                    database.addEmploi(em);
                    timeTableAvailable(em);
                    //Demande de signature d'emploi
                    //EmploiDuTemp tmpEmploi = new EmploiDuTemp(em.niveau(), "");
                    //tmpEmploi.setNomSalle(em.nomSalle());
                    //tmpEmploi.setDateDebut(em.dateDebut());
                    //tmpEmploi.setLsHoraire(new ArrayList<>());
                    //tmpEmploi.setLsMatieres(new ArrayList<>());
                    //askSignatureEmploi(em);
                }
                //qDebug()<<"onReadyRead (LAST_EMPLOI_FLAG): "<<em.niveau()<<em.lsMatieres().size();
            } else if (flag.equals(Constant.SIGNATURE_EMPLOI_FLAG)) {
                EmploiDuTemp em = readEmploi();
                byte[] signature = new byte[999999];
                try {
                    String string = null;
                    while (string.equals("")){
                        string = inputStream.readLine();
                        signature = string.getBytes();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
                //em.setNomSalle(emploi().nomSalle());
                //emit signatuereAvaible(em.niveau(), emploi().nomSalle(), em.dateDebut(), em.dateFin(), signature);
                for(SocketListener s: lsListeners)
                    s.signatureAvailable(em.niveau(), em.nomSalle(), em.dateDebut(), em.dateFin(), signature);
                quit();
            } else {
                Log.i("DEBUG", "onReadyRead unknow FLAG = " + flag);
            }
            if (objectName.equals(Constant.TEMP_SOCKET_NAME)) {
                try {
                    socket.close();
                    break;
                } catch (IOException e) {
                    e.printStackTrace();
                    break;
                }
            }
        }
    }

    void quit(){
        try {
            interrupt();
            socket.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    byte[] toBytes(String str){
        return str.getBytes();
    }

    //Lecture d'emploi du temps

    EmploiDuTemp readEmploi(){
        EmploiDuTemp em = new EmploiDuTemp("", "");
        String niveau = readLine(),
                nbreMH = readLine(),
                nomSalle = readLine(),
                semestre = readLine(), lieuCreation="";
        VDate debut = VDate.fromString(readLine()),
                fin = VDate.fromString(readLine()),
                dateCreation = VDate.fromString(readLine());
        lieuCreation = readLine();

        int nbreMatiere = 0, nbreHoraire= 0;
        try {
            nbreMatiere =  Integer.valueOf(nbreMH.split(":")[0]);
            nbreHoraire = Integer.valueOf(nbreMH.split(":")[1]);
        } catch (IndexOutOfBoundsException | NumberFormatException e) {
            Log.i("MYDEBUG", "Error at Student socket :"+e.getMessage());
        }
        //Recuperation des matieres
        List<Matiere> lsMatiere = new ArrayList<>();
        for(int i=0, j= nbreMatiere; i<j; i++){
            String typeMatiere = readLine();
            String nomMatire = readLine();
            String textNumCase = readLine();
            String gradeEnseignat = readLine();
            String nomEnseignant = readLine();
            String prenomEnseignat = readLine();
            String couleur = readLine();
            String commentaire = readLine();
            int numroCase = -1;
            try {
                numroCase = Integer.valueOf(textNumCase);
            } catch (NumberFormatException e) {
                e.printStackTrace();
            }
            Matiere matiere = new Matiere(nomMatire);
            matiere.setColor(VColor.fromString(couleur));
            matiere.setType(typeMatiere);
            Teacher t = new Teacher(nomEnseignant, prenomEnseignat, gradeEnseignat, "");
            matiere.setTeacher(t);
            matiere.setNumeroCase(numroCase);
            lsMatiere.add(matiere);

            /*Log.i("MYDEBUG", "Type de matiere ="+typeMatiere+
                    "\nNom de matiere= "+nomMatire +
                    "\nNum case = "+textNumCase+" | "+numroCase+
                    "\nTeacher=" +gradeEnseignat+" "+nomEnseignant+" "+prenomEnseignat+
                    "\nCouleur = "+couleur+"("+matiere.color().red()+", "+matiere.color().green()+", "+matiere.color().blue()+")\n"+
                    "\nComment= "+commentaire);*/
        }em.setLsMatieres(lsMatiere);
        //Recueration de la list des horaires
        List<Horaire> lsHoraires = new ArrayList<>();
        for(int i=0; i<nbreHoraire; i++){
            Horaire h = Horaire.fromString(readLine()+"\n"+readLine());
            lsHoraires.add(h);
            //Log.i("MYDEBUG", "Horaire "+h.toString() + " numCase = "+h.numeroCase()+"\n");
        }em.setLsHoraire(lsHoraires);

        nomSalle = nomSalle.equals("SALLE_VIDE") ? "" : nomSalle;
        semestre = semestre.equals("SEMESTRE VIDE") ? "" : semestre;

        em.setNiveau(niveau);
        em.setNomSalle(nomSalle);
        em.setSemestre(semestre);
        em.setDateDebut(debut);
        em.setLieuDeCreation(lieuCreation);
        em.setNomSalle(nomSalle);
        return em;
    }

    String readLine(){
        try {
            String str = inputStream.readLine();
            return str == null ? "" : str;
        } catch (IOException e) {
            e.printStackTrace();
        }
        return "";
    }

    public void onDisconnected()
    {
        connected= false;
        for(SocketListener l: lsListeners)
            l.status(Constant.OFFLINE_ICON);
    }

    EmploiDuTemp emploi(){
        return m_emploi;
    }

    public void setEmploi(EmploiDuTemp newEmploi){
        m_emploi = newEmploi;
    }

    //Demande l'emploi au serveur
    public void askEmploi(VDate dateDebut, String niveau) throws IOException {
        EmploiDuTemp em = new EmploiDuTemp(niveau, "");
        em.setDateDebut(dateDebut);
        StudentSocket socket = new StudentSocket();
        socket.setDatabase(database);
        socket.setObjectName(Constant.SIGNATURE_SOCKET_NAME);
        setEmploi(em);
        socket.start();
        //outputStream.write(FLAG + GET_EMPLOI_FLAG);
    }

    //Demande au serveur la signature de l'emploi de temps
    public void askSignatureEmploi(EmploiDuTemp em)
    {
        if(!connected)
            return;
        Log.i("MYDEBUG", "Demande de signarure pour l'emploi de "+ emploi().nomSalle());
        StudentSocket socket = new StudentSocket();
        socket.setDatabase(database);
        socket.setObjectName(Constant.ASK_TIMETABLE_SOCKET_NAME);
        setEmploi(em);
        socket.start();
    }

    public boolean isConnected(){
        return connected && socket.isConnected();
    }

    @Override
    public void timeTableAvailable(EmploiDuTemp em) {
        for (SocketListener l: lsListeners)
            l.timeTableAvailable(em);
    }

    @Override
    public void status(String state) {
        for (SocketListener l: lsListeners)
            l.status(state);
    }

    @Override
    public void signatureAvailable(String level, String nomSalle, VDate dateDebut, VDate dateFin, byte[] signature) {
        for (SocketListener l: lsListeners)
            l.signatureAvailable(level, nomSalle, dateDebut, dateFin, signature);
    }

    @Override
    public void run() {
        while (true){
            tryToConnect();
            onReadyRead();
        }
    }
}

File: Etudiant/controler/studentsocket.cpp
#include "studentsocket.h"

StudentSocket::StudentSocket(QObject *parent) : QTcpSocket(parent), connected(false)
{
    static int nbreConnexion = 0;
    database = new LocalDatabase(DATABASENAME, QString("CONNEXION %1").arg(nbreConnexion++)); connected = false;
    //setReadBufferSize(90000);
    connect(this, &QTcpSocket::connected, this, &StudentSocket::onConnected);
    connect(this, &QTcpSocket::disconnected, this, &StudentSocket::onDisconnected);
    connect(this, &QTcpSocket::readyRead, this, &StudentSocket::onReadyRead);
    connect(this, &QTcpSocket::stateChanged, this, &StudentSocket::onStateChange);
    connect(this, SIGNAL(error(QAbstractSocket::SocketError)), this, SLOT(onErrorOccurred(QAbstractSocket::SocketError)));
    tryToConnect();
}

StudentSocket::StudentSocket()
{
    static int nbreConnexion = -1;
    database = new LocalDatabase(DATABASENAME, QString("CONNEXION %1").arg(nbreConnexion--)); connected = false;
    connect(this, &QTcpSocket::readyRead, this, &StudentSocket::onReadyRead);
}

void StudentSocket::onConnected()
{
    connected= true;
    QByteArray data;
    QTextStream stream(&data, QIODevice::WriteOnly);
    stream<<FLAG; Qt::endl(stream);
    stream<<GET_LAST_EMPLOI_FLAG; Qt::endl(stream);
    write(data);
}

void StudentSocket::onDisconnected()
{
    connected= false;
    qDebug()<<"StudentSocket::onDisconnected()";
    tryToConnect();
}

//Reception de données depuis le serveur

void StudentSocket::onReadyRead()
{
    qDebug()<<QThread::currentThread();
    emit status(tr("Téléchargement d'emploi du temps en cours ..."));
    QByteArray data;
    while(!atEnd()){
        data+= read(sizeof (quint16));
        if(data.size() >= (int)sizeof(quint16)){
            break;
        }
    }
    QTextStream stream(&data, QIODevice::ReadOnly);
    int prevSize = -1;
    int round = 0;
    while(bytesAvailable()>0 || waitForReadyRead()){
        waitForReadyRead(1000);
        //qDebug()<<"data.size() = "<<data.size();
        prevSize = data.size();
        data+= readAll();
        if(prevSize == data.size() && round++ > 3)
            break;
    }
    //qDebug()<<"Taille de données recues: "<<taille<<" et data.size = "<<data.size();
    int nbreEmbloi =0;
    while(!stream.atEnd()){
        QString flag = stream.readLine();
        flag = stream.readLine();
        qDebug()<<objectName()<<flag;
        if(flag == EMPLOI_FLAG){
            EmploieDuTemp em;
            stream>>em;
            if(database->addEmploi(em)){
                qDebug()<<"//Demande de signature d'emploi";
                if(em.niveau().isEmpty()){
                    qDebug()<<"Nom de l'emploi vide"<<em.niveau()<<em.lsMatieres().size()<<em.lsHoraire().size();
                    continue;
                }
                EmploieDuTemp tmpEmploi(em.niveau());
                tmpEmploi.setNomSalle(em.nomSalle());
                tmpEmploi.setDateDebut(em.dateDebut());
                tmpEmploi.setLsHoraire(QList<Horaire>());
                tmpEmploi.setLsMatieres(QList<Matiere>());
                askSignatureEmploi(em);
                QThread::currentThread()->sleep(1);
                nbreEmbloi++;
            }
            //qDebug()<<"StudentSocket::onReadyRead: EMPLOI_FLAG "<<em.niveau()<<em.lsMatieres().size();
        }else if(flag == LAST_EMPLOI_FLAG){
            nbreEmbloi++;
            EmploieDuTemp em;
            stream>>em;
            if(em.lsMatieres().size()>0){
                database->addEmploi(em);
                emit emploisAvailables();
                emit emploiAvailable(em.dateDebut(), em.niveau());
                //Demande de signature d'emploi
                if(em.niveau().isEmpty()){
                    qDebug()<<"Nom de l'emploi vide"<<em.niveau()<<em.lsMatieres().size()<<em.lsHoraire().size();
                    continue;
                }
                EmploieDuTemp tmpEmploi(em.niveau());
                tmpEmploi.setNomSalle(em.nomSalle());
                tmpEmploi.setDateDebut(em.dateDebut());
                tmpEmploi.setLsHoraire(QList<Horaire>());
                tmpEmploi.setLsMatieres(QList<Matiere>());
                //askSignatureEmploi(em);
            }
            //emit emploisAvailables(em);
            //qDebug()<<"StudentSocket::onReadyRead (LAST_EMPLOI_FLAG): "<<em.niveau()<<em.lsMatieres().size();
        }else if(flag == SIGNATURE_EMPLOI_FLAG){
            EmploieDuTemp em;
            QByteArray signature;
            stream>>em;
            stream>>signature;
            em.setNomSalle(emploi().nomSalle());
            emit signatuereAvaible(em.niveau(), emploi().nomSalle(), em.dateDebut(), em.dateFin(), signature);
            qDebug()<<"StudentSocket::onReadyRead"<<SIGNATURE_EMPLOI_FLAG<<signature.size()<<
                      em.niveau()<<em.nomSalle();
        }
        else{
            qDebug()<<"StudentSocket::onReadyRead IGNORED FLAG = "<<bytesAvailable()<<objectName()<<flag;
        }
    }
    if(nbreEmbloi>0)
        emit emploisAvailables();
    emit status(tr("Connecté"));
    if(bytesAvailable() > 0){
        qDebug()<<"Données restantes = "<<bytesAvailable();
    }
    if(objectName() == TEMP_SOCKET_NAME){
        close();
        deleteLater();
    }
}

//
void StudentSocket::onStateChange(SocketState stat)
{
    switch (stat) {
    case QTcpSocket::ConnectingState:
        emit status(("Tentative de connexion en cours ...")); break;
    case QTcpSocket::ConnectedState:
        emit status(tr("Connecté")); break;
    case QTcpSocket::UnconnectedState:
        emit status(tr("Non connecté"));
    default:
        break;
    }
}

void StudentSocket::onErrorOccurred(SocketError error)
{
    switch (error) {
    case QTcpSocket::ConnectionRefusedError:
        emit status(tr("Connexion réfusée")); break;
    case QTcpSocket::HostNotFoundError:
        emit status(tr("Erreur de connexion: Adresse du serveur non trouvé")); break;
    case QTcpSocket::SocketTimeoutError:
        if(!isConnected())
            emit status(tr("Temps de connexion expiré"));
        break;
    case QTcpSocket::RemoteHostClosedError:
        emit status(tr("Erreur de connexion: le serveur à fermer la connexion")); break;
    default: break;
    }
    auto timer = new QTimer();
    connect(timer, &QTimer::timeout, this, &StudentSocket::tryToConnect);
    connect(timer, &QTimer::timeout, timer, &QTimer::deleteLater);
    timer->start(5000);
}

const EmploieDuTemp &StudentSocket::emploi() const
{
    return m_emploi;
}

void StudentSocket::setEmploi(const EmploieDuTemp &newEmploi)
{
    m_emploi = newEmploi;
}

void StudentSocket::tryToConnect()
{
    if(connected) emit status(tr("Connecté"));
    if(state() == ConnectingState || state() == HostLookupState || connected == true) return;
    auto pr = database->getPrametre();
    connectToHost(pr.getHostName(), pr.getPortName());
}

//Demande l'emploi au serveur
void StudentSocket::askEmploi(QDate dateDebut, QString niveau)
{
    QByteArray data;
    QTextStream stream(&data, QIODevice::WriteOnly);
    EmploieDuTemp em(niveau); em.setDateDebut(dateDebut);
    stream<<FLAG; Qt::endl(stream);
    stream<<GET_EMPLOI_FLAG; Qt::endl(stream);
    stream<<em;
    write(data);
    qDebug()<<"StudentSocket::askEmploi"<<em.dateDebut()<<em.dateFin();
}

//Demande au serveur la signature de l'emploi de temps
void StudentSocket::askSignatureEmploi(const EmploieDuTemp &em)
{
    if(!connected)
        return;
    //qDebug()<<"Demande de signarure pour l'emploi de "<<emploi().nomSalle();
    auto tmpSock = new StudentSocket();
    tmpSock->setEmploi(em);
    connect(tmpSock, &StudentSocket::status, this, &StudentSocket::status);
    connect(tmpSock, &StudentSocket::signatuereAvaible, this, &StudentSocket::signatuereAvaible);
    tmpSock->setObjectName(TEMP_SOCKET_NAME);//pour detruire a la fin de lecture de données

    connect(tmpSock, &QTcpSocket::connected, tmpSock, [tmpSock](){
        QByteArray data;
        QTextStream stream(&data, QIODevice::WriteOnly);
        stream<<FLAG; Qt::endl(stream)<<GET_SIGNATURE_EMPLOI_FLAG;
        Qt::endl(stream)<<tmpSock->emploi();
        tmpSock->write(data);
        //qDebug()<<"Demande de signarure pour l'emploi de "<<tmpSock->emploi().nomSalle()<<tmpSock->emploi().niveau();
    });
    auto thread = new QThread();
    tmpSock->setParent(nullptr);
    tmpSock->moveToThread(thread);
    connect(thread, &QThread::started, tmpSock, [tmpSock](){
        tmpSock->tryToConnect();// Se connecter pour telecharger la signature

    });
    connect(tmpSock, &QTcpSocket::destroyed, thread, [thread](){
        thread->exit();
        if(thread->wait()){
            thread->deleteLater();
        }
    });
    thread->start();
}

bool StudentSocket::isConnected() const
{
    return connected;
}
